wikimedia/wikipedia-ios
# Push notifications subscription flow

**Phabricator:** https://phabricator.wikimedia.org/T287305

### Notes
This PR is meant to lay the groundwork for getting push notifications to appear in the app. It involves subscribing the device token from the notifications framework to our echo backend, as well as asking users for notifications permissions.

Also note the changes to `NotificationSettingsViewController` are meant to be rather temporary and are just a happy path to get to an enabled permissions and subscription state. Fine-tuning to it will be made when https://phabricator.wikimedia.org/T288688 is tackled.

It's probably easiest to review this commit-by-commit. The commit messages are hopefully clear enough.

After review, please wait for #4010 to be reviewed and merged. Then update this with feature/notifications and merge.

### Test Steps
1. First update [this line](https://github.com/wikimedia/wikipedia-ios/blob/T288297/WMF%20Framework/Configuration.swift#L47) with `return Configuration.staging(options: [.betaCluster])`. We're going to temporarily force the Experimental scheme to point to the beta cluster (the sandbox flow so far only pushes to the Experimental bundle ID).
2. Create two accounts on the [beta cluster](https://en.wikipedia.beta.wmflabs.org/wiki/Main_Page). Log in on desktop with the first one. Go into preferences and make sure all web and apps column checkmarks are checked. Make an edit to any article.
3. Fresh install this app on a device running the Experimental scheme.
4. Go to Settings > Notifications.
5. ~~Proxy here to watch calls, you will reference device token posted here in later step.~~ Log in to your first account if needed here, then toggle ON and allow OS push permissions. ~~Make a note of `providertoken` value and `topic` value in call made to `action=echopushsubscriptions&command=create`.~~
6. Background the app. Now on desktop, log in to your 2nd account you made in step 2. Go to the article that your first account edited, tap into history, and thank the first account for that edit.
7. ~~Temporary step because sandbox flow is [broken again](https://phabricator.wikimedia.org/T287897#7302176): Go to push notification utility and send a visible using the device token and topic noted in step 5~~. Actually this step shouldn't be necessary anymore. The sandbox flow should be back in place now without the need to trigger via the push utility.
8. Your device should display a "checkEchoV1" push notification.
